From: Dongmao Zhang <dmzhang@suse.com>
Date: Tue, 10 Dec 2013 10:52:22 -0600
Subject: dlm: set zero linger time on sctp socket
Patch-mainline: Queued in subsystem maintainer repository
Git-repo: git://git.kernel.org/pub/scm/linux/kernel/git/teigland/linux-dlm.git
Git-commit: ece35848c1847cdf3dd07954578d3e99238ebbae
References: bnc#787843

The recovery time for a failed node was taking a long
time because the failed node could not perform the full
shutdown process.  Removing the linger time speeds this
up.  The dlm does not care what happens to messages to
or from the failed node.

Signed-off-by: Dongmao Zhang <dmzhang@suse.com>
Signed-off-by: David Teigland <teigland@redhat.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 fs/dlm/lowcomms.c |    8 ++++++++
 1 file changed, 8 insertions(+)
--- a/fs/dlm/lowcomms.c
+++ b/fs/dlm/lowcomms.c
@@ -462,6 +462,7 @@ static void process_sctp_notification(st
 				      struct msghdr *msg, char *buf)
 {
 	union sctp_notification *sn = (union sctp_notification *)buf;
+	struct linger linger;
 
 	if (sn->sn_header.sn_type == SCTP_ASSOC_CHANGE) {
 		switch (sn->sn_assoc_change.sac_state) {
@@ -539,6 +540,13 @@ static void process_sctp_notification(st
 			}
 			add_sock(new_con->sock, new_con);
 
+			linger.l_onoff = 1;
+			linger.l_linger = 0;
+			ret = kernel_setsockopt(new_con->sock, SOL_SOCKET, SO_LINGER,
+						(char *)&linger, sizeof(linger));
+			if (ret < 0)
+				log_print("set socket option SO_LINGER failed");
+
 			log_print("connecting to %d sctp association %d",
 				 nodeid, (int)sn->sn_assoc_change.sac_assoc_id);
 
